src/Polly.Core/CircuitBreaker/Health/RollingHealthMetrics.cs

```diff
@@ -0,0 +1,72 @@
+namespace Polly.CircuitBreaker.Health;
+
+internal sealed class RollingHealthMetrics : HealthMetrics
+{
+    private readonly TimeSpan _samplingDuration;
+    private readonly TimeSpan _windowDuration;
+    private readonly Queue<HealthWindow> _windows;
+
+    private HealthWindow? _currentWindow;
+
+    public RollingHealthMetrics(TimeSpan samplingDuration, short numberOfWindows, TimeProvider timeProvider)
+        : base(timeProvider)
+    {
+        _samplingDuration = samplingDuration;
+        _windowDuration = TimeSpan.FromTicks(_samplingDuration.Ticks / numberOfWindows);
+        _windows = new Queue<HealthWindow>();
+    }
+
+    public override void IncrementSuccess() => UpdateCurrentWindow().Successes++;
+
+    public override void IncrementFailure() => UpdateCurrentWindow().Failures++;
+
+    public override void Reset()
+    {
+        _currentWindow = null;
+        _windows.Clear();
+    }
+
+    public override HealthInfo GetHealthInfo()
+    {
+        UpdateCurrentWindow();
+
+        var successes = 0;
+        var failures = 0;
+        foreach (var window in _windows)
+        {
+            successes += window.Successes;
+            failures += window.Failures;
+        }
+
+        return HealthInfo.Create(successes, failures);
+    }
+
+    private HealthWindow UpdateCurrentWindow()
+    {
+        var now = TimeProvider.UtcNow;
+        if (_currentWindow == null || now - _currentWindow.StartedAt >= _windowDuration)
+        {
+            _currentWindow = new()
+            {
+                StartedAt = now
+            };
+            _windows.Enqueue(_currentWindow);
+        }
+
+        while (now - _windows.Peek().StartedAt >= _samplingDuration)
+        {
+            _windows.Dequeue();
+        }
+
+        return _currentWindow;
+    }
+
+    private class HealthWindow
```

Comment:
```suggestion
    private sealed class HealthWindow
```

Reply:
I believe .NET 8 will have analyzer for this so I won't forget it all the time then :)

Reply:
I thought we had one already - maybe it's one of the built-in SDK ones that we haven't turned on? I'm sure I've been told off about this in other non-8 code bases... 🤔

Reply:
I have turned on CA1852, however, it doesn't really trigger any warnings. Strange. Maybe on .NET 8 it will start working.